### PR TITLE
Fixed #28103 -- Added quarter extract and exposed it through quarter lookup.

### PR DESCRIPTION
**Pull request by atombrella**
https://code.djangoproject.com/ticket/28103

### CODE REVIEW

**Comment by atombrella** [comment]
buildbot, test on oracle.

**Comment by atombrella** [comment]
buildbot, test on oracle

File: django/db/backends/sqlite3/base.py [hunk 1]
```diff
@@ -2,8 +2,10 @@
 SQLite3 backend for the sqlite3 module in the standard library.
 """
 import decimal
+import math
 import re
 import warnings
+
```
Comment (charettes) [inline django/db/backends/sqlite3/base.py]: Avoid unnecessary whitespace changes, isort will complain here as well.

Comment (timgraham) [inline django/db/backends/sqlite3/base.py]: This looks unexpected per isort.

File: docs/ref/models/querysets.txt [hunk 2]
```diff
@@ -2788,6 +2788,28 @@ When :setting:`USE_TZ` is ``True``, datetime fields are converted to the
 current time zone before filtering. This requires :ref:`time zone definitions
 in the database <database-time-zone-definitions>`.
 
+.. fieldlookup:: quarter
+
+``quarter``
+~~~~~~~~~~~
+
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: `.. versionadded:: 2.0`

File: tests/db_functions/test_datetime.py [hunk 3]
```diff
@@ -413,9 +438,6 @@ def test_trunc_func(self):
         with self.assertRaisesMessage(ValueError, "Cannot truncate TimeField 'start_time' to DateTimeField"):
             list(DTModel.objects.annotate(truncated=Trunc('start_time', 'month')))
 
-        with self.assertRaisesMessage(ValueError, "Cannot truncate DateField 'start_date' to DateTimeField"):
```
Comment (timgraham) [inline tests/db_functions/test_datetime.py]: It's not obvious what this removal has to do with this patch.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by atombrella** [comment]
buildbot, test on oracle.

**Comment by atombrella** [comment]
buildbot, test on oracle.

File: django/db/backends/mysql/operations.py [hunk 4]
```diff
@@ -39,6 +39,10 @@ def date_trunc_sql(self, lookup_type, field_name):
         if lookup_type in fields:
             format_str = fields[lookup_type]
             return "CAST(DATE_FORMAT(%s, '%s') AS DATE)" % (field_name, format_str)
+        elif lookup_type == 'quarter':
+            return "MAKEDATE(YEAR(%s), 1) + INTERVAL QUARTER(%s) QUARTER - INTERVAL 1 QUARTER" % (
```
Comment (atombrella) [inline django/db/backends/mysql/operations.py]: @adamchainz maybe you know a more succinct way to do this?

File: django/db/backends/mysql/operations.py [hunk 5]
```diff
@@ -64,6 +68,12 @@ def datetime_trunc_sql(self, lookup_type, field_name, tzname):
         fields = ['year', 'month', 'day', 'hour', 'minute', 'second']
         format = ('%%Y-', '%%m', '-%%d', ' %%H:', '%%i', ':%%s')  # Use double percents to escape.
         format_def = ('0000-', '01', '-01', ' 00:', '00', ':00')
+        if lookup_type == 'quarter':
+            fmt = (
+                "CAST(DATE_FORMAT(MAKEDATE(YEAR({}), 1) + INTERVAL QUARTER({}) QUARTER - " +
+                "INTERVAL 1 QUARTER, '{}') AS DATETIME)"
+            )
+            return fmt.format(field_name, field_name, '%%Y-%%m-01 00:00:00')
```
Comment (adamchainz) [inline django/db/backends/mysql/operations.py]: Put the format string inside `fmt` because it's not really a parameter since it's always the same thing?

File: django/db/backends/mysql/operations.py (hunk 4 above)
Comment (adamchainz) [inline django/db/backends/mysql/operations.py]: can't see anything obvious after refamiliarizing myself with the [date/time functions](https://mariadb.com/kb/en/mariadb/date-and-time-functions/)

**Comment by atombrella** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
Are the boundary conditions (January, December, etc.) tested sufficiently here?

**Comment by atombrella** [comment]
buildbot, test on oracle.

**Comment by atombrella** [comment]
@timgraham I added a few more dates to the test suite. I guess they were missing.

**Comment by atombrella** [comment]
buildbot, test on oracle.

**Comment by atombrella** [comment]
buildbot, test on oracle

**Comment by atombrella** [comment]
buildbot, test on oracle.

File: django/db/backends/mysql/operations.py [hunk 6]
```diff
@@ -64,6 +68,12 @@ def datetime_trunc_sql(self, lookup_type, field_name, tzname):
         fields = ['year', 'month', 'day', 'hour', 'minute', 'second']
         format = ('%%Y-', '%%m', '-%%d', ' %%H:', '%%i', ':%%s')  # Use double percents to escape.
         format_def = ('0000-', '01', '-01', ' 00:', '00', ':00')
+        if lookup_type == 'quarter':
+            fmt = (
```
Comment (timgraham) [inline django/db/backends/mysql/operations.py]: Perhaps this would be more succinct:
```
return (
    '....{field_name}'
).format(field_name=field_name)
```

File: docs/ref/models/querysets.txt [hunk 7]
```diff
@@ -2830,6 +2830,30 @@ When :setting:`USE_TZ` is ``True``, datetime fields are converted to the
...
+.. versionadded:: 2.0
+
+For date and datetime fields, a 'quarter of the year' math. Allows chaining
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: math or match (as above)?

File: docs/releases/2.0.txt [hunk 8]
```diff
@@ -218,6 +218,14 @@ Models
   number of rows fetched by the Python database client when streaming results
   from the database. For databases that don't support server-side cursors, it
   controls the number of results Django fetches from the database adapter.
+* Added the :class:`~django.db.models.functions.datetime.ExtractQuarter`
```
Comment (timgraham) [inline docs/releases/2.0.txt]: add blank line above

File: docs/releases/2.0.txt [hunk 9]
```diff
@@ -218,6 +218,14 @@ Models
...
+  function to extract the quarter from :class:`~django.db.models.DateField` and
+  :class:`~django.db.models.DateTimeField` and exposed it through the
```
Comment (timgraham) [inline docs/releases/2.0.txt]: , and

File: docs/releases/2.0.txt [hunk 10]
```diff
@@ -218,6 +218,14 @@ Models
...
+  :lookup:`quarter` lookup.
+
+* Added the :class:`~django.db.models.functions.datetime.TruncQuarter`
+  function to truncate :class:`~django.db.models.DateField` and
+  :class:`~django.db.models.DateTimeField` to the date of a quarter.
```
Comment (timgraham) [inline docs/releases/2.0.txt]: to the first day of a quarter

File: docs/ref/models/database-functions.txt [hunk 11]
```diff
@@ -428,6 +429,12 @@ Usage example::
 
     .. attribute:: lookup_name = 'week'
 
+.. class:: ExtractQuarter(expression, tzinfo=None, **extra)
```
Comment (timgraham) [inline docs/ref/models/database-functions.txt]: Add to the examples below? I guess we're also missing ExtractWeek in there. Could be added in a separate commit and backported.

File: django/db/backends/mysql/operations.py (hunk 6 above)
Comment (adamchainz) [inline django/db/backends/mysql/operations.py]: +1 always better to eliminate a temporary variable

File: docs/ref/models/database-functions.txt (hunk 11 above)
Comment (atombrella) [inline docs/ref/models/database-functions.txt]: Yep. I can also submit a PR for the missing week examples.
